In [129]:
import pandas as pd

In [152]:
df = pd.read_csv('vacancies_merged_prizn (2) (1).csv', sep=',')
df

,Unnamed: 0,profession,profession_full,vacancy_description,date_published,payment_from,payment_to,experience,employment,city,...,company_description,vacancy_count,staff_count,catalogues,source,dms,ai,work_format,education,1с
0,0,Начальник отдела,Руководитель отдела сопровождения внутренней и...,"Компания ""Санкт-Петербургская биржа"" Вам предс...",2026-03-25,NaN,NaN,NaN,Полная занятость,Москва,...,NaN,NaN,NaN,NaN,scrapping,1,0,0,0,0
1,1,Java,Java (Senior)( ТЕХНОЛОГИИ ОТРАСЛЕВОЙ ТРАНСФОРМ...,"Компания ""ТЕХНОЛОГИИ ОТРАСЛЕВОЙ ТРАНСФОРМАЦИИ""...",2026-03-27,NaN,NaN,Более 6 лет,Полная занятость,Москва,...,NaN,NaN,NaN,NaN,scrapping,1,0,0,0,0
2,2,Специалист,Middle+ AQA специалист,"Компания ""Компания Ритейл Сервис"" Мы - аккреди...",2026-03-26,140000.0,180000.0,NaN,Полная занятость,Барнаул,...,NaN,NaN,NaN,NaN,scrapping,0,0,2,0,0
3,3,Технолог,Технолог( АКРИХИН ),"Компания ""АКРИХИН"" Обязанности: Обеспечивать ...",2026-03-25,NaN,NaN,NaN,Полная занятость,Старая Купавна,...,NaN,NaN,NaN,NaN,scrapping,1,0,0,2,0
4,4,Project manager,Project Manager Game Dev( Beresnev Games ),"Компания ""Beresnev Games"" Мы в поиске Project ...",2026-03-26,NaN,NaN,NaN,Полная занятость,Москва,...,NaN,NaN,NaN,NaN,scrapping,0,0,2,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
30862,30862,Практикант / стажёр Документовед,Практикант / стажёр Документовед,ГНЦ РФ АО «ВНИИНМ» разрабатывает конструкционн...,2026-04-08,0.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,"Российская государственная корпорация, объедин...",3.0,менее 50,"[{'id': 1, 'title': 'Административная работа, ...",api,1,0,0,0,0
30863,30863,Технический специалист/ Системный администратор,Технический специалист/ Системный администратор,"Обязанности:\n• Установка, обновление и админи...",2026-04-08,80000.0,100000.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,"Наша Школа - это пространство возможностей, тв...",1.0,менее 50,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,0,0,0,0,0
30864,30864,Стажер в IT отдел,Стажер в IT отдел,Здесь занимаются поддержкой IT-платформ и их р...,2026-04-08,0.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,Ведущий российский бренд безалкогольных напитк...,0.0,более 5000,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,1,0,1,0,0
30865,30865,Сотрудник технической поддержки,Сотрудник технической поддержки,"Добрый день!\nМы компания Lasertech, занимаемс...",2026-04-07,80000.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,Lasertech — один из ведущих игроков на рынке п...,1.0,менее 50,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,0,0,0,1,1


In [153]:
import numpy as np

In [154]:
#переиспользуем код коллег
df['payment_from_clean'] = df['payment_from']
df['payment_to_clean'] = df['payment_to']

# если payment_to меньше 1000 & payment_from > 1000, то payment_to * 1000
mask = (df['payment_to'] < 1000) & (df['payment_from'] > 1000)
df.loc[mask, 'payment_to_clean'] = df.loc[mask, 'payment_to'] * 1000

# если оба < 1000, то payment_from * 1000 + payment_to & payment_to = null
mask = (df['payment_from'] < 1000) & (df['payment_to'] < 1000)
df.loc[mask, 'payment_from_clean'] = df.loc[mask, 'payment_from'] * 1000
df.loc[mask, 'payment_to_clean'] = np.nan

# 0 = null
df.loc[df['payment_from_clean'] < 10, 'payment_from_clean'] = np.nan
df.loc[df['payment_to_clean'] < 10, 'payment_to_clean'] = np.nan

In [155]:
df

,Unnamed: 0,profession,profession_full,vacancy_description,date_published,payment_from,payment_to,experience,employment,city,...,staff_count,catalogues,source,dms,ai,work_format,education,1с,payment_from_clean,payment_to_clean
0,0,Начальник отдела,Руководитель отдела сопровождения внутренней и...,"Компания ""Санкт-Петербургская биржа"" Вам предс...",2026-03-25,NaN,NaN,NaN,Полная занятость,Москва,...,NaN,NaN,scrapping,1,0,0,0,0,NaN,NaN
1,1,Java,Java (Senior)( ТЕХНОЛОГИИ ОТРАСЛЕВОЙ ТРАНСФОРМ...,"Компания ""ТЕХНОЛОГИИ ОТРАСЛЕВОЙ ТРАНСФОРМАЦИИ""...",2026-03-27,NaN,NaN,Более 6 лет,Полная занятость,Москва,...,NaN,NaN,scrapping,1,0,0,0,0,NaN,NaN
2,2,Специалист,Middle+ AQA специалист,"Компания ""Компания Ритейл Сервис"" Мы - аккреди...",2026-03-26,140000.0,180000.0,NaN,Полная занятость,Барнаул,...,NaN,NaN,scrapping,0,0,2,0,0,140000.0,180000.0
3,3,Технолог,Технолог( АКРИХИН ),"Компания ""АКРИХИН"" Обязанности: Обеспечивать ...",2026-03-25,NaN,NaN,NaN,Полная занятость,Старая Купавна,...,NaN,NaN,scrapping,1,0,0,2,0,NaN,NaN
4,4,Project manager,Project Manager Game Dev( Beresnev Games ),"Компания ""Beresnev Games"" Мы в поиске Project ...",2026-03-26,NaN,NaN,NaN,Полная занятость,Москва,...,NaN,NaN,scrapping,0,0,2,0,0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
30862,30862,Практикант / стажёр Документовед,Практикант / стажёр Документовед,ГНЦ РФ АО «ВНИИНМ» разрабатывает конструкционн...,2026-04-08,0.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,менее 50,"[{'id': 1, 'title': 'Административная работа, ...",api,1,0,0,0,0,NaN,NaN
30863,30863,Технический специалист/ Системный администратор,Технический специалист/ Системный администратор,"Обязанности:\n• Установка, обновление и админи...",2026-04-08,80000.0,100000.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,менее 50,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,0,0,0,0,0,80000.0,100000.0
30864,30864,Стажер в IT отдел,Стажер в IT отдел,Здесь занимаются поддержкой IT-платформ и их р...,2026-04-08,0.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,более 5000,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,1,0,1,0,0,NaN,NaN
30865,30865,Сотрудник технической поддержки,Сотрудник технической поддержки,"Добрый день!\nМы компания Lasertech, занимаемс...",2026-04-07,80000.0,0.0,NaN,NaN,"{'id': 4, 'title': 'Москва', 'declension': 'в ...",...,менее 50,"[{'id': 33, 'title': 'IT, Интернет, связь, тел...",api,0,0,0,1,1,80000.0,NaN


In [156]:
df[df['ai'] == 1]['profession']

,profession
12,Менеджер
23,Senior developer
45,Cio
50,Сетевой инженер
59,Менеджер проектов
...,...
29399,Стажёр Data Science в департамент валидации и ...
29865,Начинающий инженер-исследователь в области ИИ
29867,Директор по педагогическому дизайну (языковые ...
29868,Штатный дизайнер продукта


In [157]:
df['payment_avg'] = np.where(
    (df['payment_from_clean'] > 0) & (df['payment_to_clean'] > 0),
    (df['payment_from_clean'] + df['payment_to_clean']) / 2,
    np.nan
)

# Гипотеза 1

H0: у вакансий связанных с ИИ зарплаты не отличаются от зарплат вакансий не связанных с ИИ

H1: у вакансий связанных с ИИ зарплаты отличаются от зарплат вакансий не связанных с ИИ

И в какую сторону и на сколько процентов они отличаются нам тоже интересно

In [158]:
from scipy import stats

In [159]:
df_temp = df.dropna(subset=['payment_avg'])

In [160]:
a = df_temp['payment_avg'][df_temp['ai'] == 1]
b = df_temp['payment_avg'][df_temp['ai'] == 0]

In [161]:
a.shape

(231,)

In [162]:
b.shape

(6374,)

Значений у нас везде > 30, по ЦПТ распределение выборочных средних сойдется к нормальному, поэтому ттест справится. Поставим доп параметр equal_var=False, чтобы не проверять неоднородность гипотез и перестраховаться


In [163]:
import plotly.express as px

In [164]:
fig = px.histogram(
    df,
    x='payment_avg',
    color='ai',
    marginal='box',
    barmode='overlay'
)

fig.show()

In [165]:
stats.ttest_ind(a, b, equal_var=False)

TtestResult(statistic=np.float64(8.350239940284167), pvalue=np.float64(5.844415394869378e-15), df=np.float64(235.65076268305833))

In [166]:
uplift = round(((a.mean()/b.mean()) - 1) * 100, 1)
print(f'{uplift}%')

62.4%


pvalue < 0,05 => отклоним H0, то есть есть основания полагать, что на уровне значимости 5% пользователи в группах а и б ведут себя статзначимо различно.

То есть да, вакансии в которых есть требования к владению ИИ предлагают большую зарплату, причем большую на 62%

# Гипотеза 2

график зп по времени - статзначимый ли рост и есть ли он вообще и причем тут инфляция

In [167]:
df['date_published'] = pd.to_datetime(df['date_published'])
df['month'] = df['date_published'].dt.to_period('M').astype(str)
df['day'] = df['date_published'].dt.to_period('D').astype(str)
df_trend = df.groupby('day')['payment_avg'].agg(['mean']).dropna().reset_index()
df_trend

,day,mean
0,2026-03-02,120000.000000
1,2026-03-03,140000.000000
2,2026-03-22,178333.333333
3,2026-03-23,137500.000000
4,2026-03-25,134635.430039
5,2026-03-26,129259.897829
6,2026-03-27,127453.681710
7,2026-03-28,136113.379346
8,2026-03-30,102799.300000
9,2026-03-31,71048.000000


In [168]:
fig = px.line(
    df_trend,
    x='day',
    y='mean',
    markers=True
)

fig.show()

Данных мало, не будем делать выводы

# Гипотеза 3
Линейная регрессия - можно ли предсказать зп по признакам - город, профессия, work format, education, dms

Что хочется узнать - насколько хорошо признаки объсняют зп (корреляция)

In [169]:
#переиспользуем код коллег
import ast # https://habr.com/ru/companies/piter/articles/493424/

def extract_city_name(city_value):
    if pd.isna(city_value):
        return np.nan
    if str(city_value).startswith('{'):
          city_dict = ast.literal_eval(city_value)
          return city_dict.get('title', np.nan)
    else:
          return city_value

df['city_name'] = df['city'].apply(extract_city_name)

In [170]:
#переиспользуем код коллег
city_counts = df['city_name'].value_counts()

min_count = 200
top_cities = city_counts[city_counts >= min_count].index.tolist() #убираем те, которые встречаются меньше 200 раз

df['city_grouped'] = df['city_name'].apply(
    lambda x: x if x in top_cities else 'Другие'
)

df['city_grouped'].nunique()

20

In [171]:
df = pd.get_dummies(df, columns=['city_grouped'], drop_first=True)

In [172]:
city_cols = [i for i in df.columns if i.startswith('city_grouped_')]

In [173]:
c = ['profession', 'work_format', 'education', 'dms', 'payment_avg', 'ai']
df_clean = df[c + city_cols].dropna(subset=c)

top_15 = df_clean['profession'].value_counts().nlargest(15).index
df_clean['profession_grouped'] = df_clean['profession'].apply(
    lambda x: x if x in top_15 else 'Другие')

Построим линейную регрессию методом OLS (метод наименьших квадратов), чтобы вместо ручного подсчете R2, pvalue и коэффициентов линейной регресии, можно было автоматически получить влияние (коэффициенты) каждого признака
(профессии, города, образования) на целевую переменную - зарплату, и еще и проверку их статзначимости.

In [177]:
import statsmodels.api as sm

X = pd.get_dummies(df_clean[['profession_grouped', 'work_format']], drop_first=True)

X = pd.concat([X, df_clean[city_cols[1:]]], axis=1)

X['education'] = df_clean['education']
X['dms'] = df_clean['dms']
X['ai'] = df_clean['ai']

y = df_clean['payment_avg']

X = sm.add_constant(X).astype(float)
model = sm.OLS(y, X).fit()

print(model.summary())


                            OLS Regression Results                            
Dep. Variable:            payment_avg   R-squared:                       0.277
Model:                            OLS   Adj. R-squared:                  0.273
Method:                 Least Squares   F-statistic:                     68.06
Date:                Fri, 10 Apr 2026   Prob (F-statistic):               0.00
Time:                        22:46:39   Log-Likelihood:                -82058.
No. Observations:                6605   AIC:                         1.642e+05
Df Residuals:                    6567   BIC:                         1.644e+05
Df Model:                          37                                         
Covariance Type:            nonrobust                                         
                                                                   coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------

Выводы:  
1) R2 = 0.28
наши признаки (город, профессия, ии, наличие дмс, уровень образования) объясняют 28% разброса зарплат -> то есть все таки есть какая-то статзначимая свзяь между зп и признаками, hr не ставят зп на рандом
2) У Москвы самый большой статзначимый вес из городов - то есть действительно в Москве выше зп
3) Из профессий самая большая связь у разработчика

Еще раз подтвердили статзначимое влияние ии, образования и дмс на зп